src/Libraries/Microsoft.Extensions.Http.Telemetry/Logging/Internal/HttpClientLogger.cs

```diff
@@ -36,20 +36,32 @@ internal sealed class HttpClientLogger : IHttpClientAsyncLogger
     private IHttpClientLogEnricher[] _enrichers;
 
     public HttpClientLogger(
+        ILogger<HttpClientLogger> logger,
+        IEnumerable<IHttpClientLogEnricher> enrichers,
+        IOptionsMonitor<LoggingOptions> optionsMonitor,
+        IServiceProvider serviceProvider,
```

Comment:
Even though I'm not a fan of service locator antipattern, .NET folks didn't suggest anything better: https://github.com/dotnet/runtime/issues/64427#issuecomment-1665940109

Reply:
I don't agree that injecting `IServiceProvider` is the "service locator antipattern", this is the normal DI. The "service locator" pattern was defined by Martin Fowler in https://martinfowler.com/articles/injection.html, and then debunked many times since (e.g., by Mark Seemann [in his blog](https://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/)).

My past experience tells me that it is kind of customary to put `IServiceProvider serviceProvider` as the first parameter. But I don't have any specific guidelines to back that up.

Reply:
I'm referring to this technique as a service locator pattern mainly because of https://learn.microsoft.com/en-us/previous-versions/msp-n-p/hh860410(v=pandp.51).
It's also considered a good practice to have all dependencies listed explicitly - in a constructor in that particular case.
Strictly speaking - yes, it's not a static service locator described by Martin Fowler

Reply:
I will put `serviceProvider` as a first argument, as you suggested

Reply:
You can also check [Dependency Injection Principles, Practices, and Patterns](https://learning.oreilly.com/library/view/dependency-injection-principles/9781617294730/), specifically chapter 4.1 where it's described that using DI container outside of the composition root is essentially a service locator pattern anyway.

Reply:
> It's also considered a good practice to have all dependencies listed explicitly - in a constructor in that particular case.

Yes, however there's a price to pay when passing multiple parameters in a constructor (any method).

In a general case I strongly prefer for all dependencies to be resolved in a constructor, but whether those dependencies are provided explicitly or come from a IServiceProvider I don't have a strong preference.

Reply:
Done